posixlib/src/main/resources/scala-native/sys/socket.c

```diff
@@ -76,6 +76,14 @@ int scalanative_so_rcvtimeo() { return SO_RCVTIMEO; }
 
 int scalanative_so_reuseaddr() { return SO_REUSEADDR; }
 
+int scalanative_so_reuseport() {
+#ifdef SO_REUSEPORT
+    return SO_REUSEPORT;
+#else
+    return 0;
+#endif
+}
```

Comment:
I followed the pattern of other code in this file, but is this really what we want to do? It seems like this will create a silent runtime failure in user code.

Instead, I wonder if we should not emit the method at all if `SO_REUSEPORT` is not defined. Then, it will be a linking error.

Reply:
I think this is a 'pick your poison and live with it'  choice

SO* are all positive integers.  0 will be an invalid socket option 
and will cause the C call to fail. I believe, I have not tried.
I could tomorrow, after sunrise.

-1 would make the invalid SO intention more obvious, possible with a few word comment.

You are correct, that this will give a runtime error to anyone who happens to execute
that piece of code. It it is on the critical path of the program, this could be every user,
and why did the developer not test it?   If it is on a side path, many users can get value
without encountering the error.

If it is a link time error, then the developer is hard stopped and has to address the
issue before they can gain value from the rest of the program.  If they do any testing
at all and if they bother to test return code, the failure should be obvious to them
before it ever reaches a user.  If they do not check return codes, they deserve
no mercy. 

At this point, I suggest picking the one you like so that you can get on with
the valuable work you are trying to do.

SO_REUSEPORT is relatively recent, within a few years, so it is likely
to be missing on a number of operating systems.  Does that cause a 
problem for your library?

PS: In my quick run through include files, I saw that `uv.h` used or mentioned
       SO_REUSEPORT.  Is that uv related to libuv and your efforts?

Reply:
Thanks for the commentary. I will leave it as-is since it follows the existing pattern.

I am implementing this because it is used in the test suite for the fs2-io library.
https://github.com/typelevel/fs2/blob/0e573f2ca899886af57aaea9edd7a1072b4eb00a/io/jvm/src/test/scala/fs2/io/net/tcp/SocketSuitePlatform.scala#L32

It appears to be a part of the JDK as of Java 9.
https://docs.oracle.com/javase/9/docs/api/java/net/StandardSocketOptions.html#SO_REUSEPORT

It's quite possible that this is used in libuv. At this time I am not using libuv though, if you are curious I added some commentary about that :)
https://github.com/armanbilge/epollcat#why-not-use-libuv

---

Tangential, but do you know if there's a reason why `SOL_TCP` might be missing? It is another one I have encountered.

Reply:
Missing from Scala Native or missing from C code?   

As you are probably encountering, it is pretty essential in order
to set socketoptions.   It is high enough in the pyramid almost
assuredly be present on any OS from this century (Socket Option Layer?).
It is not one of the "twidle this bit in some obscure driver" os dependent ones.

If  `SOL_TCP` is missing from C code, that probably
means that the proper .h file has not been included.  As a path we have both walked down, this can be
OS specific but is, I believe  <netinet/tcp.h>.

If it is missing in Scala, then it is an oops!    I think I used this 10 days or so ago (yup, my bad karma is
showing).

Let me know if you can find it.

Reply:
Yes, it is missing from Scala Native :) I was afraid it might be an OS-specific option, but this seemed odd since as you point it it is very essential.

In that case I might sneak it into this PR.

Reply:
A few minutes later.  Scala Native 0.5.0-S posixlib netinet/in.h has one (interesting) line
`int scalanative_tcp_nodelay() { return TCP_NODELAY; }`
 
Nothing else.   No SOL_TCP.  No SOL_UDP.

What other socket options are you using?

I can grep all the .h file to see if those two options, and possibly
others, exist elsewhere in SN.

Thought you would like the late breaking news.

Reply:
Yes, I just added `TCP_NODELAY` in https://github.com/armanbilge/epollcat/pull/6, which is where I ran into trouble with no `SOL_TCP` :)

These are the ones I am currently using. They are all present, except for `SO_REUSEPORT` and `SOL_TCP`.
https://github.com/armanbilge/epollcat/blob/84b1010a77ba823b70dc3ed89125cca4406a9f5a/net/src/main/scala/java/net/StandardSocketOptions.scala

Reply:
No SOL_TCP that I can find in any SN .scala, .c, or .h files.

If you give me a laundry list, I can implement some/most (0.4.x)
tomorrow daylight. Or do you need them sooner so it is easier 
if you do them?   Since we are trying to do rapid development,
I am trying to cut down strictly clerical work at your end.

Please advise,

Reply:
Actually, what would be very helpful is to upstream the entirety of `StandardSocketOptions` into Scala Native 0.4.x, and make sure the constants are available for each of those options. Then I can remove it from my project.

Here is the full list of options as of JDK 17.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/StandardSocketOptions.html

Reply:
OK, I will look at that.  Thank you for the URL.

 ` SocketOption<NetworkInterface> IP_MULTICAST_IF` may be hard to do in economic time.
I think I have looked at the NetworkInterface mumble before, actually several times. when I was
looking for a way to see if IPv6 was enabled on a system.

Reply:
I just did a whirlwind tour of `SocketOption` and checked what was available from Apache Harmony.

We may need to talk in a PM on Discord or such.  Java 8 is a class with static methods. I think this becomes
a Scala `object`.   Java 17 is an Interface, again with static methods.   IIUC an Interface translates to a Scala
trait. I believe that traits can be instantiated (new) as though they were a class. However I think static
methods must be in an `object`.   `SocketOptions` are also an Interface. It looks like in Scala it would be
an abstract class. I have to figure out the class to implement the abstract class.  

Have you figured all this stuff out already or are you using, by necessity, workarounds?

You have set a good problem for me.   Good night.

Reply:
Lee, thanks for all the investigations! Much appreciated, your rigor is duly noted.

Indeed, this is a bit of an unusual case. However I think putting static methods in the `object` is the correct idiom. Perhaps we can check Scala.js for some inspiration: we won't find these specific APIs, but some other API that is a Java `interface` with static methods.

I am currently using this:
https://github.com/armanbilge/epollcat/blob/84b1010a77ba823b70dc3ed89125cca4406a9f5a/net/src/main/scala/java/net/StandardSocketOptions.scala

It seems to be working perfectly well on JDK 17. I have not tested an older JDK yet.

Talk soon, good night!